src/lib/x509/x509_ext.cpp

```diff
@@ -847,6 +856,518 @@ void TNAuthList::decode_inner(const std::vector<uint8_t>& in) {
+   // in a partially used byte
+   uint8_t unused_bits = host % 8;
+
+   bool octets_match = true;
+   for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets - 1); i++) {
+      if(min[i] != max[i]) {
+         octets_match = false;
+         break;
+      }
+   }
+
+   // we only partially use this octet, and the used part has to match for prefix encoding
+   bool used_bits_match = true;
+   // if we would discard all octets we don't actually have a partially used octet, so this doesn't matter
+   if(discarded_octets < version_octets - 1) {
+      uint8_t shifted_min = (min[version_octets - 1 - discarded_octets] >> unused_bits);
+      uint8_t shifted_max = (max[version_octets - 1 - discarded_octets] >> unused_bits);
+      used_bits_match = (shifted_min == shifted_max);
+   }
+
+   // both the full octets and the partially used one match
+   if(octets_match && used_bits_match) {
+      // at this point the range can be encoded as a prefix
+      std::vector<uint8_t> prefix;
+
+      prefix.push_back(unused_bits);
+      for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets); i++) {
+         prefix.push_back(min[i]);
+      }
+
+      into.add_object(ASN1_Type::BitString, ASN1_Class::Universal, prefix);
+   } else {
+      uint8_t discarded_octets_min = zeros / 8;
+      uint8_t unused_bits_min = zeros % 8;
+
+      uint8_t discarded_octets_max = ones / 8;
+      uint8_t unused_bits_max = ones % 8;
+
+      // compress the max address by setting unused bits to 0, for the min address these are already 0
+      if(unused_bits_max != 0) {
+         max[version_octets - 1 - discarded_octets_max] >>= unused_bits_max;
+         max[version_octets - 1 - discarded_octets_max] <<= unused_bits_max;
+      }
+
+      std::vector<uint8_t> compressed_min;
+      std::vector<uint8_t> compressed_max;
+
+      // construct the address as a byte sequence of the unused bits followed by the compressed address
+      compressed_min.push_back(unused_bits_min);
+      for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets_min); i++) {
+         compressed_min.push_back(min[i]);
+      }
+
+      compressed_max.push_back(unused_bits_max);
+      for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets_max); i++) {
+         compressed_max.push_back(max[i]);
+      }
+
+      into.start_sequence()
+         .add_object(ASN1_Type::BitString, ASN1_Class::Universal, compressed_min)
+         .add_object(ASN1_Type::BitString, ASN1_Class::Universal, compressed_max)
+         .end_cons();
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressOrRange<V>::decode_from(Botan::BER_Decoder& from) {
+   const uint8_t version_octets = static_cast<uint8_t>(V);
+
+   BER_Object obj = from.get_next_object();
+   ASN1_Type type_tag = obj.type_tag();
+
+   // this can either be a prefix or a single address
+   if(type_tag == ASN1_Type::BitString) {
```

Comment:
This function similar to `ASBlocks::ASIdOrRange::decode_from` can be made much simpler by using `peek_next_object` to decide if you are in the `IPAddress` or `IPAddressRange` case. Then you can use the builtin functionality for decoding bit strings, parsing sequences etc without having to replicate logic here.